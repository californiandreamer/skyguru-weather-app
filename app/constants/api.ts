// TODO: move to env
export const API_KEY = 'fafbc890379de1e204d9cfc570657337'

export const BASE_URL = 'https://api.openweathermap.org/data/2.5/'
