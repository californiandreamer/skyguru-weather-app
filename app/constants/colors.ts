export const transparent = 'transparent'
export const skyBlue = '#26B3F0'
export const darkOrange = '#DAB866'
export const darkViolet = '#0D0411'
export const transparentGray = 'rgba(51, 51, 51, 0.8)'
export const violet = '#240024'
export const white = '#EEE'
export const gray = '#878787'
