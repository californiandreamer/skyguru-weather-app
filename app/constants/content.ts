import { NoConnectionIcon } from 'app/assets'

export const noConnectionContent = {
  illustration: NoConnectionIcon,
  text: 'It seems you are not connected to the internet. Please check your connection and try again',
}

export const locationPermissionContent = {
  illustration: null,
  text: 'It looks like you denied location permission. Please give access to the geolocation in your phone setting for this app',
}
