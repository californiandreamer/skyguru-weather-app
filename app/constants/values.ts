// Gaps
export const littleGap = 5
export const smallGap = 10
export const gap = 20
export const largeGap = 40

// Fonts
export const littleFontSize = 16
export const smallFontSize = 20
export const defaultFontSize = 24
export const largeFontSize = 72
export const defaultLineHeight = 32

// Objects
export const defaultIconSize = 20
export const weatherIconSize = 36
export const largeIconSize = 72

// Defaults
export const resultsQuantity = 5
export const fadeTiming = 500
export const pressOpacity = 0.8
export const daysQuantity = 6
export const units = 'metric'
