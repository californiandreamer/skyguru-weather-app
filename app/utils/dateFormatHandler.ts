export const dateFormatHandler = (
  timestamp: number,
  type: 'day' | 'time'
): string => {
  const converted = new Date(timestamp * 1000)
  const time = converted.toTimeString().split(' ')[0]
  const day = converted.toLocaleString('pl-PL').split(',')[0]

  switch (type) {
    case 'time':
      return time
    case 'day':
      return day
  }
}
