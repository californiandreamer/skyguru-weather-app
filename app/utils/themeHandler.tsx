import {
  DayEclipseImage,
  DayMountainBottomImage,
  DayMountainMiddleImage,
  DayMountainTopImage,
  EveningEclipseImage,
  EveningMountainBottomImage,
  EveningMountainMiddleImage,
  EveningMountainTopImage,
  MoonImage,
  NightEclipseImage,
  NightMountainBottomImage,
  NightMountainMiddleImage,
  NightMountainTopImage,
  SunImage,
} from 'app/assets'
import { darkOrange, darkViolet, skyBlue } from 'app/constants/colors'
import { DayPartT } from 'app/store/types/theme'
import React from 'react'
import { SvgProps } from 'react-native-svg'

export type ObjectTypeT = 'mountains' | 'sky' | 'sun' | 'eclipse'
export type SunTypeT = 'sun' | 'moon'

export type ThemeHandlerT = React.FC<SvgProps> | null

export const themeHandler = (value: DayPartT | SunTypeT, type: ObjectTypeT) => {
  const dayMountains = [
    DayMountainTopImage,
    DayMountainMiddleImage,
    DayMountainBottomImage,
  ]

  const nightMountains = [
    NightMountainTopImage,
    NightMountainMiddleImage,
    NightMountainBottomImage,
  ]

  const eveningMountains = [
    EveningMountainTopImage,
    EveningMountainMiddleImage,
    EveningMountainBottomImage,
  ]

  const dayEclipse = DayEclipseImage
  const nightEclipse = NightEclipseImage
  const eveningEclipse = EveningEclipseImage

  const sun = SunImage
  const moon = MoonImage

  if (type === 'mountains') {
    switch (value) {
      case 'night':
        return nightMountains
      case 'morning':
        return eveningMountains
      case 'day':
        return dayMountains
      case 'evening':
        return eveningMountains
      default:
        return dayMountains
    }
  } else if (type === 'sun') {
    switch (value) {
      case 'day':
        return sun
      case 'night':
        return moon
      default:
        return sun
    }
  } else if (type === 'eclipse') {
    switch (value) {
      case 'night':
        return nightEclipse
      case 'morning':
        return eveningEclipse
      case 'day':
        return dayEclipse
      case 'evening':
        return eveningEclipse
      default:
        return dayEclipse
    }
  } else if (type === 'sky') {
    switch (value) {
      case 'night':
        return darkViolet
      case 'morning':
        return darkOrange
      case 'day':
        return skyBlue
      case 'evening':
        return darkOrange
      default:
        return skyBlue
    }
  }
}
