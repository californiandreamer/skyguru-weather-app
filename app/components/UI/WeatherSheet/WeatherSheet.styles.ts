import { gap, smallGap } from 'app/constants/values'
import { StyleSheet } from 'react-native'

export default StyleSheet.create({
  container: {
    paddingHorizontal: 40,
  },
  item: {
    width: '100%',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingVertical: smallGap,
  },
  itemWrapper: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  itemIcon: {
    paddingLeft: gap,
  },
})
