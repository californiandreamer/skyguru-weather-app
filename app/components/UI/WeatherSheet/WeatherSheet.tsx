import StyledText from 'app/components/HOC/Text'
import { daysQuantity, pressOpacity } from 'app/constants/values'
import { IDayWeather } from 'app/models'
import { showWeatherInfo } from 'app/store/actions/weatherInfo'
import { RootState } from 'app/store/reducers/rootReducer'
import { dayOfWeekHandler } from 'app/utils/dayOfWeekHandler'
import { weatherFormatter } from 'app/utils/weatherFormatter'
import { weatherIconHandler } from 'app/utils/weatherIconHandler'
import React, { useEffect, useState } from 'react'
import { FlatList, TouchableOpacity, View } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'

import styles from './WeatherSheet.styles'

type WeatherSheetT = {
  onTouchMoveDown: (locationY: number) => void
  onTouchMoveEnd: () => void
}

const WeatherSheet: React.FC<WeatherSheetT> = ({
  onTouchMoveDown,
  onTouchMoveEnd,
}) => {
  const dispatch = useDispatch()

  const [data, setData] = useState<IDayWeather[] | null>(null)
  const slicedData = data?.slice(0, daysQuantity)

  const { futureWeather } = useSelector((state: RootState) => {
    return state.futureWeather
  })

  const handleWeatherData = () => {
    if (futureWeather) {
      setData(futureWeather.daily)
    }
  }

  useEffect(() => {
    handleWeatherData()
  }, [futureWeather])

  const renderWeatherItem = (item: IDayWeather) => (
    <TouchableOpacity
      style={styles.item}
      key={item.dt}
      activeOpacity={pressOpacity}
      accessibilityLabel="daily item"
      onPress={() => {
        dispatch(showWeatherInfo(item))
      }}
    >
      <StyledText>{dayOfWeekHandler(item.dt)}</StyledText>
      <View style={styles.itemWrapper}>
        <StyledText>
          {weatherFormatter(item.temp.day, 'temperature')}
        </StyledText>
        <View style={styles.itemIcon}>
          {weatherIconHandler(
            item.weather.map((weatherItem) => weatherItem.description)[0]
          )}
        </View>
      </View>
    </TouchableOpacity>
  )

  const renderWeatherItems = () => (
    <FlatList
      data={slicedData}
      renderItem={({ item }) => renderWeatherItem(item)}
      keyExtractor={(item) => item.dt.toString()}
      // TODO: implement pull-to-refresh logic
      // bounces={true}
      // bouncesZoom={true}
      // refreshing={false}
      // showsVerticalScrollIndicator={false}
      // onRefresh={() => {
      //   // TODO: add common data fetch saga
      // }}
      // onTouchMove={({ nativeEvent }) => {
      //   if (nativeEvent.locationY > 40 && nativeEvent.locationY < 180) {
      //     onTouchMoveDown(nativeEvent.locationY)
      //   }
      // }}
      // onTouchEnd={() => onTouchMoveEnd()}
    />
  )

  return <View style={styles.container}>{renderWeatherItems()}</View>
}

export default WeatherSheet
