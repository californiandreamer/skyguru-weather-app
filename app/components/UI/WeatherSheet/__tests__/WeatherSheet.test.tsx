import createSagaMiddleware from '@redux-saga/core'
import { render } from '@testing-library/react-native'
import { daysQuantity } from 'app/constants/values'
import { dayOfWeekHandler } from 'app/utils/dayOfWeekHandler'
import React from 'react'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import createMockStore from 'redux-mock-store'

import WeatherSheet from '../WeatherSheet'

jest.useRealTimers()

describe('WeatherSheet component', () => {
  const dayTimestamp = 86400
  const timestamp = Date.now() / 1000
  const initialState = {
    futureWeather: {
      futureWeather: {
        daily: [
          {
            dt: timestamp,
            temp: { day: 15 },
            weather: [{ description: 'clear sky' }],
          },
          {
            dt: timestamp + dayTimestamp,
            temp: { day: 15 },
            weather: [{ description: 'clear sky' }],
          },
          {
            dt: timestamp + dayTimestamp * 2,
            temp: { day: 15 },
            weather: [{ description: 'clear sky' }],
          },
          {
            dt: timestamp + dayTimestamp * 3,
            temp: { day: 15 },
            weather: [{ description: 'clear sky' }],
          },
          {
            dt: timestamp + dayTimestamp * 4,
            temp: { day: 15 },
            weather: [{ description: 'clear sky' }],
          },
          {
            dt: timestamp + dayTimestamp * 5,
            temp: { day: 15 },
            weather: [{ description: 'clear sky' }],
          },
          {
            dt: timestamp + dayTimestamp * 6,
            temp: { day: 15 },
            weather: [{ description: 'clear sky' }],
          },
          {
            dt: timestamp + dayTimestamp * 7,
            temp: { day: 15 },
            weather: [{ description: 'clear sky' }],
          },
        ],
      },
    },
  }
  const saga = createSagaMiddleware()
  const middlewares = [saga]
  const mockStore = createMockStore(middlewares)

  const mock = mockStore(initialState)

  const component = (
    <Provider store={mock}>
      <WeatherSheet />
    </Provider>
  )

  it('returns correct quantity of days', () => {
    const tree = render(component)
    const { getAllByLabelText } = tree

    const result = getAllByLabelText('daily item')

    expect(result.length).toBe(daysQuantity)
  })

  it('returns correct day', () => {
    const tree = render(component)
    const { getByText } = tree

    const result = getByText(dayOfWeekHandler(timestamp))

    expect(result.props.children).toBe(dayOfWeekHandler(timestamp))
  })

  it('renders correctly', () => {
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
