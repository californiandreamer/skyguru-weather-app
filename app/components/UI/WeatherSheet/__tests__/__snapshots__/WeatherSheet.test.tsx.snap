// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherSheet component renders correctly 1`] = `
<View
  style={
    Object {
      "paddingHorizontal": 40,
    }
  }
>
  <RCTScrollView
    bounces={false}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</View>
`;
